### ISSUE TITLE
has_one through not available on unpersisted instances

### ISSUE THREAD

**Issue by zachwalton** [rails/rails#33155]
Opening a new issue since the old one was marked as stale and I didn't get to it in time (doesn't seem like I can re-open it).

https://github.com/rails/rails/issues/31975

@adamlogic confirmed that this still reproduces on current master: https://github.com/rails/rails/issues/31975#issuecomment-397656705

Note: I believe the way that this works is by associating with post_id, so if that's the case it makes sense that this would be the behavior, and I suppose this becomes an enhancement request.

`Comment` knows that it has a `Post` associated with it as evidenced by the passed assertion on L66, so is it possible to leverage that relationship to make the `through` association work?

### Steps to reproduce

Script to reproduce: https://gist.github.com/zachwalton/395412f5b7e97837e6273981dbb7faf3

### Expected behavior

`owner` should be not be nil when accessed through the `has_one through` association

### Actual behavior

`owner` is nil when accessed through the `has_one through` association

### System configuration
**Rails version**: 5.1.2

**Ruby version**: 2.4.2

**Comment by jychen7** [comment]
i believe the gist is not a normal example of has one through.

http://guides.rubyonrails.org/association_basics.html#the-has-one-through-association

it expects A has one B, B belongs to C, then A has one C through B.

but now the gist is A belongs to B, B belongs to C.

**Comment by liligga** [comment]
@jychen7 I guess you misunderstood/confused the concept of this association. Just look again at the docs you provided([HasOneThrough](http://guides.rubyonrails.org/association_basics.html#the-has-one-through-association)). It is rather `A has_one B` and `B has_one C` and thus we can write `A has_one C, through: B`

**Comment by jherdman** [comment]
I'm running into this as recent as of Rails 5.2.3.

**Comment by keoghpe** [comment]
I've just run into this issue. Has any more progress been made on this?

**Comment by rjclaasen** [comment]
This also seems to occur with `has_many through` relationships in Rails 6.1.4:

Script to reproduce: https://gist.github.com/rjclaasen/6b16aa896d622e20f9faf7737c89fc94

**Comment by chriscz** [comment]
I have a temporary patch, though there are very likely corner-cases that I've missed. I will have to test a bit more, but this can be included into your `ApplicationRecord`

https://gist.github.com/chriscz/335c6d6bbafebbd1a3c874c3b4aa3ffa

**Comment by joshuay03** [comment]
I've created minimal reproduction script for both `has_one through:` and `has_many through:` and attempted to patch it in #51114. Just gonna wait on some initial feedback there from the maintainers.

Script (swap the gem path to run against the patch):

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # issue reproduction
  gem "rails", github: "rails/rails", branch: "main"
  # patch
  # gem "rails", github: "joshuay03/rails", branch: "fix-through-target-unavailable-on-new-records"

  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :authors, force: true do |t|
    t.string :name
  end

  create_table :posts, force: true do |t|
    t.integer :author_id
  end

  create_table :comments, force: true do |t|
    t.integer :post_id
  end

  create_table :categories, force: true do |t|
    t.integer :post_id
  end
end

class Author < ActiveRecord::Base
end

class Post < ActiveRecord::Base
  belongs_to :author
  has_many :comments
  has_many :categories
end

class Comment < ActiveRecord::Base
  belongs_to :post
  has_one :author, through: :post
  has_many :categories, through: :post
end

class Category < ActiveRecord::Base
  belongs_to :post
end

class BugTest < Minitest::Test
  def test_has_one_through_with_new_parent_record
    author = Author.create!(name: "Josh")
    post = Post.new(author: author) # this needs to be a new record so the comment isn't persisted
    comment = Comment.new # this needs to be a new record to reproduce the bug
    post.comments << comment

    assert_predicate author, :persisted?
    assert_predicate post, :new_record?
    assert_predicate post.comments.first, :new_record?
    assert_equal author.name, post.comments.first.post.author.name
    assert_equal author.name, post.comments.first.author.name
  end

  def test_has_many_through_with_new_parent_record
    category_1 = Category.create!
    category_2 = Category.create!
    post = Post.new # this needs to be a new record so the comment isn't persisted
    post.categories << category_1
    post.categories << category_2
    comment = Comment.new # this needs to be a new record to reproduce the bug
    post.comments << comment

    assert_predicate category_1, :persisted?
    assert_predicate category_2, :persisted?
    assert_predicate post, :new_record?
    assert_predicate post.comments.first, :new_record?
    assert_equal [category_1, category_2], post.comments.first.post.categories.to_a
    assert_equal [category_1, category_2], post.comments.first.categories.to_a
  end
end

```

**Comment by chriscz** [comment]
@joshuay03 this is great! I didn't see it when I started implementing above. Definitely hope it gets merged, thank you!

**Comment by KerLq** [comment]
@chriscz's workaround works perfectly! Thank you :D

Hopefully this issue gets resolved asap and doesn't require a workaround 👍

**Comment by sromano** [comment]
We should reopen this issue which is still present, now that it has been reverted by https://github.com/rails/rails/pull/54280
It's an important bug that affects active record deeply.

I do not have permission to reopen. CC @adrianna-chang-shopify @matthewd  and @joshuay03 

In our app, the solution was differently than the one implemented (we are using that in our initializer as a patch in rails 8.0.2.1)
We can help to debug other proposed implementation now that we have upgraded from Rails 7 to Rails 8

```
module ActiveRecord
  module Associations
    class Association
      def load_target
        @target = find_target(async: false) if (@stale_state && stale_target?) || find_target?
        
        if !@target && set_through_target_for_new_record?
          reflections = reflection.chain
          reflections.pop
          reflections.reverse!
          @target = reflections.reduce(through_association.target) do |middle_target, reflection|
            break unless middle_target
        
            middle_target.association(reflection.send(:actual_source_reflection).name).load_target
          end
        end
        
        loaded! unless loaded?
        target
      rescue ActiveRecord::RecordNotFound
        reset
      end
      
      private
      
      def set_through_target_for_new_record?
        owner.new_record? && reflection.through_reflection? && through_association.target.present?
      end
    end
  end
end

```

**Comment by joshuay03** [comment]
Thanks for the poke. I was hoping to reattempt those changes, but I'm unsure when I'll be able to get to it. Feel free to send a PR if you've got an implementation that gets you a green build 👍🏽
